src/cpp/src/scheduler.hpp

```diff
@@ -405,6 +432,82 @@ class Scheduler {
+        }
+
+        if (device_type == ov::device::Type::INTEGRATED) {
+            used_device_mem += memory_statistics["usm_host"];
+        }
+
+        // there could be unaccounted extra memory reserved by kernels, kept
+        // in memory pools, etc
+        // therefore, add a threshold to account for this
+        float used_memory_threshold = 1.1;
+        used_device_mem *= used_memory_threshold;
+
+        // total device memory in bytes
+        auto total_device_memory = core->get_property(device, ov::intel_gpu::device_total_mem_size);
+
+        return total_device_memory - used_device_mem;
+    }
+
+    void _initialize_cache(const std::vector<SequenceGroup::Ptr>& sequence_groups) {
+        size_t seq_length_sum = 0;
+        for (auto idx = 0; idx < sequence_groups.size(); idx++) {
+            auto seq_length = sequence_groups[idx]->get_prompt_len() * m_kv_blocks_initial_multiplier;
+            auto gen_config = sequence_groups[idx]->get_sampling_parameters();
+            seq_length = std::min(seq_length, sequence_groups[idx]->get_prompt_len() + gen_config.get_max_new_tokens(sequence_groups[idx]->get_prompt_len()));
+            if (sequence_groups[idx]->get_sampling_parameters().is_beam_search()) {
+
+                seq_length *= sequence_groups[idx]->get_sampling_parameters().num_beams;
+            }
+            seq_length_sum += seq_length;
+        }
+        m_block_manager.increase_kv_blocks_number(seq_length_sum);
+        m_cache_manager->allocate_cache_if_needed(m_block_manager.get_total_number_of_kv_blocks());
+        m_dynamic_memory_allocation = true;
+    }
+
+    void _try_increase_cache() {
+        if (!m_dynamic_memory_allocation) {
+            return;
+        }
+        auto device_config = m_cache_manager->get_device_config();
+        auto device = device_config->get_device();
+        size_t current_num_of_kv_blocks = m_block_manager.get_total_number_of_kv_blocks();
+        size_t new_blocks_num = current_num_of_kv_blocks * m_cache_growth_factor;
+
+        if (device.find("GPU") == std::string::npos) {
+            m_block_manager.increase_kv_blocks_number(new_blocks_num);
+        }
+        else {
+            size_t available_gpu_memory = _get_available_gpu_memory();
+            size_t required_memory = (new_blocks_num - current_num_of_kv_blocks) * device_config->get_block_size_in_bytes();
+            if (required_memory <= available_gpu_memory) {
+                m_block_manager.increase_kv_blocks_number(new_blocks_num);
+            } else {
+                size_t possible_blocks_to_add = available_gpu_memory / device_config->get_block_size_in_bytes();
+                if (possible_blocks_to_add > 0) {
+                    m_block_manager.increase_kv_blocks_number(current_num_of_kv_blocks + possible_blocks_to_add);
+                }
+            }
+        }
+        m_cache_manager->allocate_cache_if_needed(m_block_manager.get_total_number_of_kv_blocks());
```

Comment:
`_try_increase_cache` can be called multiple times within `schedule` and it means we will reallocate KV cache and perform blocks copy several times

maybe we need to call `m_cache_manager->allocate_cache_if_needed`  only once at end of `schedule` call?
In this case we can allocate logical blocks several times, but physically KV cache is touched only once

Reply:
Yes, done.